std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,680 @@
+
+    // Allocations of size 128 will go to the third 'AlignedBlockList'
+    foreach (j; 0 .. testNum)
+    {
+        buf[j] = a.allocate(128);
+        assert(buf[j].length == 128);
+
+        // This is owned by the third 'AlignedBlockList'
+        assert(a.allocatorForSize!128.owns(buf[j]) == Ternary.yes);
+    }
+
+    // Free the memory
+    foreach (j; 0 .. testNum)
+        assert(a.deallocate(buf[j]));
+
+    // Allocations which exceed 128, will go to the 'AscendingPageAllocator*'
+    void[] b = a.allocate(256);
+    assert(b.length == 256);
+    a.deallocate(b);
+}
+
+/**
+`SharedAlignedBlockList` is the threadsafe version of `AlignedBlockList`.
+The `Allocator` template parameter must refer a shared allocator.
+Also, `ParentAllocator` must be a shared allocator, supporting `alignedAllocate`.
+*/
+shared struct SharedAlignedBlockList(Allocator, ParentAllocator, ulong theAlignment = (1 << 21))
+{
+    version (StdDdoc)
+    {
+        import std.typecons : Ternary;
+        import std.traits : hasMember;
+
+        /**
+        Returns a chunk of memory of size `n`
+        It finds the first node in the `AlignedBlockNode` list which has available memory,
+        and moves it to the front of the list.
+
+        All empty nodes which cannot return new memory, are removed from the list.
+
+        Params:
+            n = bytes to allocate
+        Returns:
+            A chunk of memory of the required length or `null` on failure or
+        */
+        static if (hasMember!(ParentAllocator, "alignedAllocate"))
+        void[] allocate(size_t n);
+
+        /**
+        Deallocates the buffer `b` given as parameter. Deallocations take place in constant
+        time, regardless of the number of nodes in the list. `b.ptr` is rounded down
+        to the nearest multiple of the `alignment` to quickly find the corresponding
+        `AlignedBlockNode`.
+
+        Params:
+            b = buffer candidate for deallocation
+        Returns:
+            `true` on success and `false` on failure
+        */
+        bool deallocate(void[] b);
+
+        /**
+        Returns `Ternary.yes` if the buffer belongs to the parent allocator and
+        `Ternary.no` otherwise.
+
+        Params:
+            b = buffer tested if owned by this allocator
+        Returns:
+            `Ternary.yes` if owned by this allocator and `Ternary.no` otherwise
+        */
+        static if (hasMember!(ParentAllocator, "owns"))
+        Ternary owns(void[] b);
+    }
+    else
+    {
+        import std.math : isPowerOf2;
+        static assert(isPowerOf2(alignment));
+        mixin AlignedBlockListImpl!true;
+    }
+}
+
+///
+@system unittest
+{
+    import std.experimental.allocator.building_blocks.region : SharedRegion;
+    import std.experimental.allocator.building_blocks.ascending_page_allocator : SharedAscendingPageAllocator;
+    import std.std.experimental.allocator.building_blocks.null_allocator : NullAllocator;
```

Comment:
std/experimental/allocator/building_blocks/aligned_block_list.d(512): Error: module `null_allocator` is in file 'std/std/experimental/allocator/building_blocks/null_allocator.d' which cannot be read
import path[0] = ../druntime/import
posix.mak:350: recipe for target 'generated/linux/debug/32/unittest/std/experimental/allocator/building_blocks/aligned_block_list.o' failed
make[1]: *** [generated/linux/debug/32/unittest/std/experimental/allocator/building_blocks/aligned_block_list.o] Error 1

-> You have a `std.std`